---
title: "Pagoo: An encapsulated OO class system for analyzing bacterial pangenomes"
author: "Ignacio Ferr√©s"
date: "April 26, 2019"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

_"One object to store them all, one object to find them, one object to query from and with ggplot2 visualize them." (Lord Sauron)_

## Introduction

This package provides an encapsulated Object Oriented class system for analyzing bacterial pangenomes. It uses the [R6](https://r6.r-lib.org/) package as backend. It was designed in order to facilitate and speed up **a lot** my projects as a microbial genomics researcher, standarizing and optimizing routine tasks of my everyday work activities.

A few years on this field made me realize that there are three things I do almost every day when working with bacterial pangenomes: 

1. Subset
2. Apply statistical methods
3. Visualize

so I wanted to have all these three tasks as light and easy for me as possible. 

This package comes with three R6 classes, each one more complex than the other, and that the more basic ones are inherited by the more complex ones. The most basic one, `PgR6`, contains basic subset methods and data manipulation functions. The second one, `PgR6M` inherits all the methods and fields from the previous one, and incorporates statistical and visualization methods. The last one, `PgR6MS`, inherits from `PgR6M` all its capabilities, and adds methods for manipulate DNA sequences. 

It is worth noting that this package is intended to facilitate the analysis of pangenomes _once the pangenome of a clade has already been obtained_, so you can reconstruct the pangenome with any of your favorite tools, and use _pagoo_ to analyze it.


### Installation 

The easiest way to install this package is using `devtools`:

```{r, eval=FALSE}
devtools::install_github('iferres/pagoo')
```

### Help

All three classes are documented. You can access R help pages as with any other function:

```{r, eval=FALSE}
library(pagoo)
?PgR6
help('PgR6MS')
```

## Example

To show the full capabilities of `pagoo` I will use the most complex class, `PgR6MS` so you can see all the fancy things you can do. The example uses toy data that comes with the package.

### Input

First of all, lets extract and decompress the example data attatched to this package:

```{r}
library(pagoo)
tgz <- system.file('extdata', 'toy_data.tar.gz', package = 'pagoo')
untar(tarfile = tgz, exdir = tempdir())
files <- list.files(path = tempdir(), full.names = TRUE, pattern = 'tsv$|fasta$')
files
```

There are 3 `.tsv` files, and 5 `.fasta` files. The `case_df.tsv` file is the main one, having the information about the membership of each gene from each organism, to each group of orthologous (clusters). `case_*_meta.tsv` files have metadata for the clusters and for the organisms. `organism*.fasta` files have the sequences for each organism:

```{r}
df <- read.table(files[1], header = TRUE, sep = '\t', quote = '')
head(df)
```
As you can see, `df` is a `data.frame` with 4 columns: the first one with the name of each gene, the second one with the organism to which each gene belongs, the third one with the cluster to which each gene was assigned in the pangenome reconstruction, and the last one with annotation metadata for each gene. Of the 4 columns, the former 3 are required, and `pagoo` will look for columns named "gene", "org", and "group". More columns are optional, and you can add as many as you want (or none) to add metadata of each gene.

The next 2 `.tsv` files contanins metadata for each cluster and for each organism, respectively, and are optional arguments. 

```{r}
#
group_meta <- read.table(files[2], header = TRUE, sep = '\t', quote = '', na.strings = '<NA>')
org_meta <- read.table(files[3], header = TRUE, sep = '\t', quote = '', na.strings = '<NA>')
head(group_meta)
head(org_meta)
```
In both cases, the columns identifying each cluster and each organism must coincide with the names provided in the fist data frame, so `pagoo` can map the metadata correctly.

As we will be working with the class `PgR6MS` in this example, DNA sequences are also required. The other 2 classes are lightweight, and don't use all this data, so keep it in mind you can work without sequences if you have limited computational resources.

```{r}
suppressMessages(library(Biostrings))
fasta_files <- files[4:8]
names(fasta_files) <- sub('[.]fasta', '', basename(fasta_files))
# A list of DNAStringSet, with organism names as list names, and DNAString names as gene names:
seqs <- lapply(fasta_files, readDNAStringSet)
```
Keep in mind that names should match with those provided in the data frames before.

### Load class

Once we have all the input data, lets load the example into a `PgR6MS` class:

```{r}
pg <- PgR6MS$new(DF = df, org_meta = org_meta, group_meta = group_meta, sequences = seqs)
pg
```
Wow. I know. A better print method will be written in the future, this is the default print method for `R6` classes. But you can see that basically there are public and private fields/methods. In private, all raw data is stored. You will not have easy access to it. Instead, you will be able to access public functions and active bindings. Active bindings are functions that behave as it were variables, you will understand it better with some examples below.

### 0. Basic fields and functions

What do you want from a pangenome? Well, let see some examples:

```{r}
# Get a panmatrix
pg$pan_matrix[, 1:5]
# Get organism information
pg$organisms
# Get cluster information
pg$clusters
# Get gene information
pg$genes
# Get sequences
pg$sequences
# Access each cluster
pg$sequences[[1]]
# What is the default core level?
pg$core_level #You can change it by assiging another value (<=100)
# Get core clusters
pg$core_clusters
# Get shell genes
pg$shell_genes
# Get cloud sequences
pg$cloud_sequences
# Get summary statitics
pg$summary_stats
```

### 1. Subset

Of course, you can subset each of the above fields as you would subset any other vector or list in R, but I wanted something better.
Lets see a little example:
```{r}
pg$pan_matrix[1:2, 1:3]
```

I want to extract those precise genes. Wouldn't be nice to have the possibility of using the same subset notation in lists as it were a matrix? Well, you can:
```{r}
pg[1:2, 1:3]$genes
```
I'm referencing the same genes in both the panmatrix and in the list above, using the same notation. You can use it with sequences also:
```{r}
pg[1:2, 1:3]$sequences
```

The default print method for `DNAStringSetList` ([Biostring](https://bioconductor.org/packages/release/bioc/html/Biostrings.html) package) is not showing what's inside each list element, but you will check it has the precise sequences we asked for if you look inside each element.

Another tedious task I found working with microbial genomics is when you run a pangenome tool over a dataset, and when you analize it you realize that one or some of the input genomes have very bad quality, or you want to remove them from downtream analysis for any other cause. I implemented `drop()` and `recover()` functions to remove or reload particular organisms from the data set:

```{r}
# Original dataset
pg$pan_matrix[, 1:5]
# I want to drop organismD because blah blah
pg$drop('organismD')
# Now is was removed from the dataset
pg$pan_matrix[, 1:5]
pg$genes[1:3]
pg$organisms
# And summary statistics are also updated (compare with pg$summary_stats called before)
pg$summary_stats
```
Now you want to get it back:
```{r}
# Any dropped organism?
pg$dropped
# You can drop/recover them by name or by index
pg$recover(4)
pg$organisms
```

Note that you don't have to reassign the result of the function to a new variable, that's [R6](https://r6.r-lib.org/articles/Introduction.html#r6-classes) reference semantics magic. Use with [caution](https://r6.r-lib.org/articles/Introduction.html#cloning-objects).


### 2. Applying statistical methods

Once you have a pangenome, you may want to get some statistics about it. I implemented some methods in `pagoo`, and take some others from widely used packages as [vegan](https://github.com/vegandevs/vegan) and [micropan](https://github.com/larssnip/micropan). You can contribute by adding more.

```{r}
# Default: vegan::vegdist(method='bray')
pg$dist()
# micropan::fluidity
pg$fluidity()
# Fitting a power law to pangenome rarefaction curves
pg$pg_power_law_fit(n.perm=100)
```

...and some others. It is worth noting that each statistic will be computed for available organisms, not taking into consideration previously dropped ones.

### 3. Visualizing results

I used [ggplot2](https://ggplot2.tidyverse.org/) as backend for vizualization methods.

```{r}
# Summary pie
pg$gg_pie()
# Classic barplot
pg$gg_barplot()
# Binmap
pg$gg_binmap()
# Pangenome and coregenome curves
pg$gg_curves()
# Customize with ggplot2
library(ggplot2)
pg$gg_curves() + 
  geom_point() + 
  facet_wrap(~category, scales = 'free_y')
```

### Recipes

I call recipes to short and easy scripts that solves a specific problem. Here I show a few to demostrate the potentialities of `pagoo`. You can help by expanding the following with more examples:

#### Plot Coregenome alignment NJ tree

One standard analysis when dealing with a pangenome is to align each core cluster, concatenate them into a super alignment, and infer a phylogeny. Since core clusters might contain in-paralogs, you need to select no more than one of them on each cluster from each organism. Also, if `core_level` (that is, the percentage of organisms a gene has to be in, to be considered core) is set below 100%, some clusters might have few organisms missing, so you would have to fill with gaps and also not miss the sequence order before concatenating. For this purpose I wrote the `core_seqs_4_phylo()` function, which return something similar to `core_sequences` field, but with above considerations.
```{r, eval=FALSE}
library(Biostrings) # DNAString xscat
library(magrittr)   # %>% (pipe operator)
library(DECIPHER)   # AlignSeqs() 
library(phangorn)   # phyDat(), dist.ml(), NJ()

lapply(pg$core_seqs_4_phylo(), AlignSeqs) %>%   # Align each core cluster
  do.call(xscat, .) %>%                         # Concatenate alignments
  as('matrix') %>%                              # Transform to matrix
  phyDat(type = 'DNA') %>%                      # Transform to phangorn's phyDat
  dist.ml() %>%                                 # Compute distance
  NJ() %>%                                      # Compute NJ
  plot()                                        # Plot it
```

#### Blastp cluster 1 against custom blast db

You can use the [rBLAST](https://github.com/mhahsler/rBLAST) package to search a specific set of sequences on a custom blast database. In the example below, sequences are first translated, and then searched against a protein blast database. The result is returned as a `data.frame` for parsing and better curation.

```{r, eval=FALSE}
devtools::install_github('mhahsler/rBLAST')
library(Biostrings)
library(rBLAST)
library(magrittr)

db_path <- 'path/to/custom/blastpdb'
db <- blast(db = db_path, type = 'blastp')
pg$sequences[[1]] %>%
  translate() %>%
  predict(db, .)
```

